src/MongoDB.Driver.Core/Core/Misc/SemanticVersion.cs

```diff
@@ -313,5 +321,159 @@ public static bool TryParse(string value, out SemanticVersion result)
+                        return -1;
+                    }
+
+                    result = _releaseCandidate.Value.CompareTo(other._releaseCandidate.Value);
+                    if (result != 0)
+                    {
+                        return result;
+                    }
+                }
+
+                if (_internalBuild != null || other._internalBuild != null)
+                {
+                    if (_internalBuild == null)
+                    {
+                        return -1;
+                    }
+                    if (other._internalBuild == null)
+                    {
+                        return 1;
+                    }
+
+                    result = _internalBuild.Value.CompareTo(other._internalBuild.Value);
+                    if (result != 0)
+                    {
+                        return result;
+                    }
+                }
+
+                if (_commitHash == null && other._commitHash == null)
+                {
+                    return 0;
+                }
+                if (_commitHash == null)
+                {
+                    return -1;
+                }
+
+                return _commitHash.CompareTo(other._commitHash);
+            }
+
+            // static methods
+            public static ServerVersion FromSemanticVersion(SemanticVersion semanticVersion)
+            {
+                var preRelease = semanticVersion.PreRelease;
+
+                if (preRelease == null)
+                {
+                    return new ServerVersion(
+                        major: semanticVersion.Major,
+                        minor: semanticVersion.Minor,
+                        patch: semanticVersion.Patch,
+                        releaseCandidate: null,
+                        internalBuild: null,
+                        commitHash: null);
+                }
+
+                var pattern = @"^((rc(?<releaseCandidate>\d+))?-?((?<internalBuild>\d+)-g(?<commitHash>[0-9a-f]{4,40}))?)$";
+                var match = Regex.Match(semanticVersion.PreRelease, pattern);
+
+                var releaseCandidateEntry = match.Groups["releaseCandidate"].Success ? match.Groups["releaseCandidate"].Value : null;
+                var internalBuildEntry = match.Groups["internalBuild"].Success ? match.Groups["internalBuild"].Value : null;
+                var commitHash = match.Groups["commitHash"].Success ? match.Groups["commitHash"].Value : null;
+
+                int? releaseCandidate = null;
+                if (releaseCandidateEntry != null && int.TryParse(releaseCandidateEntry, out int releaseCandidateParsed))
+                {
+                    releaseCandidate = releaseCandidateParsed;
+                }
+                int? internalBuild = null;
+                if (internalBuildEntry != null && int.TryParse(internalBuildEntry, out int internalBuildParsed))
+                {
+                    internalBuild = internalBuildParsed;
+                }
```

Comment:
Lines 451-467 can be simplified to:

    var serverPreReleasePattern = @"^(rc-(?<releaseCandidate>\d+)-)?((?<commitsAfterRelease>\d+)-g(?<commitHash>[0-9a-f]{4,40}))?$";
    var match = Regex.Match(semanticVersion.PreRelease, serverPreReleasePattern);
    if (!match.Success)
    {
        throw new FormatException($"Invalid server preRelease string: \"{preRelease}\".");
    }

    var releaseCandidateString = match.Groups["releaseCandidate"].Value;
    var commitsAfterReleaseString = match.Groups["commitsAfterRelease"].Value;
    var commitHash = match.Groups["commitHash"].Value;

    var releaseCandidate = releaseCandidateString == "" ? default(int?) : int.Parse(releaseCandidateString);
    var commitsAfterRelease = commitsAfterReleaseString == "" ? default(int?) : int.Parse(commitsAfterReleaseString);

Reply:
You probably need to do some more work on this class.

For example, what happens if `preRelease` is `4.4.0-beta1`?